### PR TITLE
wip

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — COMMENTED** [review]
Thanks for writing this up. I think I understand it a bit more now but there are still a bunch of points that are unclear to me:

1. Why do we need the unknown version in speedy? I get that values coming from the ledger API are not normalized but we need to traverse those fully anyway to validate them so it seems very possible to combine that with encoding them as the correct version. I can find a bunch of usages in tests but no other usages. Am I missing something here? I totally understand that it is convenient to have this for tests but I don’t really like the idea of having logic just for tests in the very core of our interpreter.

2. If we don’t need the unknown version in speedy, I don’t understand why we need it in `Value`. Shouldn’t we be able to chose a specific version when going from protobuf -> `Value` (apologies, I can’t find the function that does the decoding here atm)?

3. I’d like to confirm that I understand the problem this is solving fully so let me give my explanation and you correct me if I’m wrong.

    1. With #8062 we don’t have any of those issues. Speedy produces values with the record id so when encoding we can simply omit it for some of them.
    2. This PR changes this. Speedy now does not always include record ids. If we were to always omit them in speedy, then we have to recover them when encoding older values. This could be expensive (question on that below) so we don’t want to do this.
    3. By having two version of the record in speedy, we can make sure that if we need it (i.e. for older versions), speedy still includes the record constructor meaning that we don’t run into the issue where we have to recover it when encoding an older value.
    4. While slowing down encoding of older values is acceptable, figuring out if we have an older value somewhere nested inside requires traversing the value fully which we would then also have to do for newer versions so we pay the price for those as well.

    Is that correct? If it is, I don’t entirely understand the performance problem for newer values. My understanding is that encoding traverses the value fully anyway (`ValueCoder.encodeValue`) so for newer versions we pay the price for the traversal anyway. At that point checking for an older version seems easy and not problematic from a performance pov. What am I missing here?

File: daml-script/converter/src/main/scala/Converter.scala [hunk 1]
```diff
@@ -39,9 +39,8 @@ private[daml] object Converter {
   // Helper to make constructing an SRecord more convenient
   def record(ty: Identifier, fields: (String, SValue)*): SValue = {
     val fieldNames = fields.view.map { case (n, _) => Name.assertFromString(n) }.to(ImmArray)
-    val args =
-      new util.ArrayList[SValue](fields.map({ case (_, v) => v }).asJava)
-    SRecord(ty, fieldNames, args)
+    val args = new util.ArrayList[SValue](fields.map({ case (_, v) => v }).asJava)
+    SRecord10(ty, fieldNames, args)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-script/converter/src/main/scala/Converter.scala:43]: You are right. This is just for the prototype.
Good to keep track of that.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala [hunk 2]
```diff
@@ -474,18 +474,16 @@ private[lf] object Pretty {
           x match {
             case SBConsMany(n) => text(s"$$consMany[$n]")
             case SBCons => text(s"$$cons")
-            case SBRecCon(id, fields) =>
+            case SBRecCon10(id, fields) =>
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:477]: We need to pretty print both.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 3]
```diff
@@ -84,13 +83,7 @@ object SExpr {
   /** Reference to a builtin function */
   final case class SEBuiltin(b: SBuiltin) extends SExprAtomic {
     def lookupValue(machine: Machine): SValue = {
-      /* special case for nullary record constructors */
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:87]: I think we do not need it, but (if we can) we should do it in another PR.

[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:87]: same response. ;)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 4]
```diff
@@ -137,9 +136,28 @@ object SValue {
...
+  }
+
+  object SRecord {
+    def unapply(record: SRecord): Some[util.ArrayList[SValue]] = Some(record.values)
+  }
+
   @SuppressWarnings(Array("org.wartremover.warts.ArrayEquals"))
-  final case class SRecord(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue])
-      extends SValue
+  final case class SRecord10(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue])
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:149]: Probably not a good name. 
It is for LF 1.6, 1.7, 1.8, 1.10 and 1.11. 
SRecord12 is for LF 1.12 or latter.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala [hunk 5]
```diff
@@ -66,7 +66,11 @@ object Ast {
   final case class EPrimLit(value: PrimLit) extends Expr
 
   /** Record construction. */
-  final case class ERecCon(tycon: TypeConApp, fields: ImmArray[(FieldName, Expr)]) extends Expr
+  final case class ERecCon(
+      tycon: TypeConApp,
+      fieldNames: ImmArray[FieldName],
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala:71]: Again it is an optim, I should not have put there.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala [hunk 6]
```diff
@@ -245,10 +250,59 @@ object Value extends CidContainer1[Value] {
...
-  final case class ValueRecord[+Cid](
-      tycon: Option[Identifier],
+  sealed abstract class ValueRecord[+Cid] extends Value[Cid] {
+    def mbTyCon: Option[Identifier]
+    def fields: ImmArray[(Option[Name], Value[Cid])]
+    def fieldValues: ImmArray[Value[Cid]]
+    def update[Cid2 >: Cid](values: ImmArray[Value[Cid2]]): ValueRecord[Cid2]
+  }
+
+  final case class ValueRecord0[+Cid](
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala:260]: Yes.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala [hunk 7]
```diff
@@ -212,7 +167,17 @@ class ValueCoderSpec
       ValueCoder.decodeVersionedValue(ValueCoder.CidDecoder, encodedSentOverWire)
     )
 
-    decodedSentOverWire.value shouldEqual value0
+    decodedSentOverWire.value shouldEqual normalizedValue
     decodedSentOverWire.version shouldEqual version
   }
+
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueCoderSpec.scala:173]: I will take that out.

**Comment by remyhaemmerle-da** [comment]
>     1. Why do we need the unknown version in speedy?

We don't.

> I get that values coming from the ledger API are not normalized but we need to traverse those fully anyway to validate them so it seems very possible to combine that with encoding them as the correct version. I can find a bunch of usages in tests but no other usages. Am I missing something here? I totally understand that it is convenient to have this for tests but I don’t really like the idea of having logic just for tests in the very core of our interpreter.

>     2. If we don’t need the unknown version in speedy, I don’t understand why we need it in `Value`. Shouldn’t we be able to chose a specific version when going from protobuf -> `Value` (apologies, I can’t find the function that does the decoding here atm)?

The proto from the ledger api is different 
- ledger-api/grpc-definitions/com/daml/ledger/api/v1/value.proto
- The converter is `com.daml.ledger.api.validation.ValueValidator`. We do not want to change the ledger API and force the user to use different message for the two kind of records, furthermore at this point we do not have access to the packages and cannot infer the kind of record should be use. 

> 
>     3. I’d like to confirm that I understand the problem this is solving fully so let me give my explanation and you correct me if I’m wrong.
>        
>        1. With #8062 we don’t have any of those issues. Speedy produces values with the record id so when encoding we can simply omit it for some of them.

      - Yes. but with #8062 have other issues: 
          - with #8062, we need a type inference step in `Machine#importValue` here we can omit it. 
          - here the validation (`isReplayedBy`) is simpler. The two value should match exactly, in #8062 we need to take care of the case there is not type information in the "recorded" transaction. 
          - here we do not need the type info at all in speedy for (`SRecord12`)
      

>        2. This PR changes this. Speedy now does not always include record ids. If we were to always omit them in speedy, then we have to recover them when encoding older values. This could be expensive (question on that below) so we don’t want to do this.

   - Yes. 

>        3. By having two version of the record in speedy, we can make sure that if we need it (i.e. for older versions), speedy still includes the record constructor meaning that we don’t run into the issue where we have to recover it when encoding an older value.

  - Yes. 

>        4. While slowing down encoding of older values is acceptable, figuring out if we have an older value somewhere nested inside requires traversing the value fully which we would then also have to do for newer versions so we pay the price for those as well.
>        
>        Is that correct? If it is, I don’t entirely understand the performance problem for newer values. My understanding is that encoding traverses the value fully anyway (`ValueCoder.encodeValue`) so for newer versions we pay the price for the traversal anyway. At that point checking for an older version seems easy and not problematic from a performance pov. What am I missing here?

The point is that we do not need type inference at all (expcet in the ledger API for the verbose mode)
      

Imagine we have 2 package the first one compile in `1.10` and the second compile in `1.12`

 ```    
       module Mod10 {
          record Template10  = {x: Text} 
         
          template (this : Template) =  {
             precondition True,
             signatories (Nil @Party),
             observers (Nil @Party),
             agreement "Agreement for Mod:Template",
             choices { }
          }; 
       }
```
```       
       module Mod12 {
          record Template12  = {x10: Mod10:Template10} 
         
          template (this : Template2) =  {
             precondition True,
             signatories (Nil @Party),
             observers (Nil @Party),
             agreement "Agreement for Mod:Template",
             choices {
               choice @nonConsuming CreateTemplate10 (self) (x: Unit) : Unit 
               to
                 ubind
                   created: ContractId Mod10:Template10 <- create @Mod10:Template10 (Mod12:Template12 {x10} sefl,
                 in 
                   pure @Unit ()
             }
          };
         
       }
```       

 
We create  in a first transaction a contract based on the `Mod:Template12`.

With #8062, the record `Mod10:Template10` nested in the contract  is serialized without type info. so to exercise `Mod:Template12:CreateTemplate10` which will create a node of version 10 , we will need to run so type inference somewhere to recover this info.

Here we simply do not need because `Mod10:Template10` record are store with all the info.

**Comment by cocreature** [comment]
>   The converter is com.daml.ledger.api.validation.ValueValidator. We do not want to change the ledger API and force the user to use different message for the two kind of records, furthermore at this point we do not have access to the packages and cannot infer the kind of record should be use.

Ah I missed that because `ValueRecord` is now a synonym for the unknown version so this didn’t change. Fully agree on not changing the ledger API. What I don’t understand is why we cannot get access to the package here. We typecheck the value later which does need the package. Why can’t we get access to the package immediately in `ValueValidator`?


> With #8062, the record `Mod10:Template10` nested in the contract is serialized without type info. so to exercise `Mod:Template12:CreateTemplate10` which will create a node of version 10 , we will need to run so type inference somewhere to recover this info.
>
> Here we simply do not need because Mod10:Template10 record are store with all the info.

Thanks for the example, that’s super helpful! I understand the tradeoff between different types and adding type inference. What I don’t quite get is why type inference incurs a significant performance overhead for new values. It seems like we can do this in the same pass as encoding the transaction (`Transaction/ValueCoder` I believe) and just push the type info down as we traverse the values. First, is there a reason why this doesn’t work? e.g., is there a similar issue where we cannot get type info in the transaction coder to make this work? Assuming it does work, am I just underestimating the costs of pushing down the type info and this still has a significant performance hit?

Just to be clear, I’m not advocating for anything here, only trying to understand tradeoffs. I do like the idea of only doing the inference for encoding in one place (the transaction coder) and not having to carry several different representations for everything through the whole stack but assuming we can make that work at all, not having any inference seems clearly better from a perf pov (whether significant or not might be better judged by benchmarks) and while perhaps a bit more code probably simpler.

**Pull request closed without merging**